### PR TITLE
add timeout for IsGCP

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fix: https://github.com/istio/istio/issues/39950

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [x] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/bootstrap/platform/gcp.go [hunk 1]
```diff
@@ -122,7 +123,24 @@ func IsGCP() bool {
 		// Assume this is running on GCP if GCP project env variable is set.
 		return true
 	}
-	return metadata.OnGCE()
+
+	timeC := time.After(time.Millisecond * 100)
+	ch := make(chan bool)
+	defer close(ch)
+	go func() {
+		ch <- metadata.OnGCE()
```
Comment (hzxuzhonghu) [inline pkg/bootstrap/platform/gcp.go:131]: this is very dangerous, sending to a closed channel can cause panic?

Comment (hzxuzhonghu) [reply pkg/bootstrap/platform/gcp.go:131]: The best fix should be in google's lib.

Here you can not close the `ch`

File: pkg/bootstrap/platform/gcp.go [hunk 2]
```diff
@@ -122,7 +123,24 @@ func IsGCP() bool {
...
+	}()
+
+	for {
```
Comment (hzxuzhonghu) [inline pkg/bootstrap/platform/gcp.go:134]: rm

**Review by howardjohn — COMMENTED** [review]
https://github.com/istio/istio/blob/e2ce47e2cb85306bcf57160ae3c4a9026cb28018/pkg/bootstrap/platform/discovery.go#L64 is already handling timeouts. Maybe we are in-appropriately calling this outside of the function?

File: pkg/bootstrap/platform/gcp.go [hunk 3]
```diff
@@ -122,7 +123,23 @@ func IsGCP() bool {
 		// Assume this is running on GCP if GCP project env variable is set.
 		return true
 	}
-	return metadata.OnGCE()
+
+	timeC := time.After(time.Millisecond * 100)
+	ch := make(chan bool)
+	go func() {
+		ch <- metadata.OnGCE()
```
Comment (hzxuzhonghu) [inline pkg/bootstrap/platform/gcp.go:130]: This can be blocked when real timeout happen

Comment (Random-Liu) [reply pkg/bootstrap/platform/gcp.go:130]: Just FYI. In k8s, it just check the local file `/sys/class/dmi/id/product_name` to decide whether it is on GCE or not
https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/gcp/metadata.go#L137

Comment (zirain) [reply pkg/bootstrap/platform/gcp.go:130]: ok, I will revisit that, thanks.

**Comment by howardjohn** [comment]
that may not be accessible from a pod. not sure

**Pull request closed without merging**